### PR TITLE
Several allocation improvements for HttpClient/ManagedHandler

### PR DESCRIPTION
**Pull request by stephentoub**
Some of these changes are specific to ManagedHandler, while others are applicable to HttpClient in general.

ManagedHandler-specific:
- We currently write out a host header with a port number by converting the port to a string.  This changes it to just write out the digits directly without the intermediate string.
- HttpRequestHeaders is currently always allocated, even if it's never accessed by the developer code.
- Parsing the response very likely yields at least once, and currently it's a helper that's called from the main SendAsync.  It's easy to combine them and avoid the extra allocations from the extra async layer.
- We're going to need to revisit the connection pooling anyway, but for now I was seeing allocations related to the ConcurrentDictionary show up.  ConcurrentyDictionary is optimized for many-reads-fewer-writes, but this usage is entirely writes.  CD just does striped locking for writes internally, so I replaced it with a simple striped-lock-HashSet structure that doesn't allocate on every add.
- We're currently incurring an array enumerator allocation on every header being written out.  We can avoid that with a cast.

HttpClient-general:
- Currently GetStream/String/ByteArrayAsync all incur a task allocation for getting the response stream.  But in most cases, the stream is available promptly (it's the reads on the stream that are likely to be async).  We can employ ValueTask to minimize the need for this allocation.
- HttpRequestHeaders has 9 reference fields dedicated to specific lazily-allocated collection properties, and HttpResponseHeaders has 5.  That's 112 bytes of fields allocated per request on 64-bit that are often unneeded.  This puts them into a lazily-allocated array to make them more pay for play: they're a bit more expensive when used (the space for the array's overhead plus a cast on each use), but we avoid having to pay for the allocations for the fields when they're not used.

Both:
- Both HttpRequestHeader and HttpResponseHeader currently promptly allocate another wrapped HttpGeneralHeaders object, but this is only really necessary in the implementation when various strongly-typed properties are accessed.  We can make it lazily-allocated and more pay-for-play.  This changes the headers to be lazily-allocated, and then modifies the ManagedHandler to avoid accessing them unless necessary.

cc: @davidsh, @geoffkizer

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/ByteArrayContent.cs [hunk 1]
```diff
@@ -62,9 +62,13 @@ protected internal override bool TryComputeLength(out long length)
...
 
-        protected override Task<Stream> CreateContentReadStreamAsync()
-        {
-            return Task.FromResult<Stream>(new MemoryStream(_content, _offset, _count, writable: false));
-        }
+        protected override Task<Stream> CreateContentReadStreamAsync() =>
+            Task.FromResult<Stream>(CreateMemoryStreamForByteArray());
+
+        internal override ValueTask<Stream> CreateContentReadStreamValueAsync() =>
+            GetType() == typeof(ByteArrayContent) ? new ValueTask<Stream>(CreateMemoryStreamForByteArray()) :
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/ByteArrayContent.cs]: Why do we need the GetType() check?

File: src/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs [hunk 2]
```diff
@@ -53,5 +53,9 @@ private static string Encode(string data)
             // Escape spaces as '+'.
             return Uri.EscapeDataString(data).Replace("%20", "+");
         }
+
+        internal override ValueTask<Stream> CreateContentReadStreamValueAsync() =>
+            GetType() == typeof(FormUrlEncodedContent) ? new ValueTask<Stream>(CreateMemoryStreamForByteArray()) :
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs]: Same as above

**Comment by geoffkizer** [comment]
Generally LGTM; one question above.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpRequestHeaders.cs [hunk 3]
```diff
@@ -15,73 +15,49 @@ public sealed class HttpRequestHeaders : HttpHeaders
...
+        private const int AcceptSlot = 0;
+        private const int ExpectSlot = 1;
+        private const int IfMatchSlot = 2;
+        private const int IfNoneMatchSlot = 3;
+        private const int TransferEncodingSlot = 4;
+        private const int UserAgentSlot = 5;
+        private const int AcceptCharsetSlot = 6;
+        private const int AcceptEncodingSlot = 7;
+        private const int AcceptLanguageSlot = 8;
+        private const int NumCollectionsSlots = 9;
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpRequestHeaders.cs]: nit:  why aren't these in alphabetical ordering as they are in numerical ordering?

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs [hunk 4]
```diff
@@ -15,28 +15,35 @@ public sealed class HttpResponseHeaders : HttpHeaders
         private static readonly Dictionary<string, HttpHeaderParser> s_parserStore = CreateParserStore();
         private static readonly HashSet<string> s_invalidHeaders = CreateInvalidHeaders();
 
+        private const int AcceptRangesSlot = 0;
+        private const int WwwAuthenticateSlot = 1;
+        private const int ProxyAuthenticateSlot = 2;
+        private const int ServerSlot = 3;
+        private const int VarySlot = 4;
+        private const int NumCollectionsSlots = 5;
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: nit: same comment as above regarding ordering.

File: src/System.Net.Http/src/System/Net/Http/ByteArrayContent.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/ByteArrayContent.cs]: Derived types may override CreateContentReadStreamAsync. If they do and we don't delegate to it, callers of this method will get the wrong behavior. As there's no fast way to ask "does this instance's method X == this type's method X", the best we can do is base it on the type itself. We could implement expensive reflection-based logic and cache the results, but that's unlikely to be worth it. If the runtime provided such functionality, we could switch to that in the future.

File: src/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/FormUrlEncodedContent.cs]: Ditto :)

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpRequestHeaders.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpRequestHeaders.cs]: I just kept the same ordering the fields were in (they weren't alphabetical, either). I can change it though.

File: src/System.Net.Http/src/System/Net/Http/ByteArrayContent.cs (hunk 1 above)
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/ByteArrayContent.cs]: Makes sense, thanks.  Can we add a comment?

**Pull request merged**